File: Turn/Turn/GameScene.m
//
//  GameScene.m
//  Turn
//
//  Created by Johannes Wärn on 09/12/14.
//  Copyright (c) 2014 Johannes Wärn. All rights reserved.
//

#import "GameScene.h"
#import "Player.h"
#import "Enemy.h"

@interface GameScene () <SKPhysicsContactDelegate>

@property (nonatomic) Player *player;

@property (nonatomic) SKLabelNode *scoreNode;
@property (nonatomic) int score;

@end

@implementation GameScene

-(void)didMoveToView:(SKView *)view {
    typeof(self) __weak weakSelf = self;
    
    self.physicsWorld.gravity = CGVectorMake(0, 0);
    self.physicsWorld.contactDelegate = self;
    
    self.physicsBody = [SKPhysicsBody bodyWithEdgeLoopFromRect:self.frame];
    self.physicsBody.friction = 0.0;
    self.physicsBody.collisionBitMask = 1 << 2;
    
    [self runAction:[SKAction repeatActionForever:[SKAction sequence:@[
                                                                       [SKAction waitForDuration:5.5 withRange:16.0],
                                                                       [SKAction runBlock:^{ [weakSelf addEnemy]; }]
                                                                       ]]]];
    
    [self setupScoreLabel];
    [self setupPlayer];
}

- (SKShapeNode *)addEnemy
{
    SKShapeNode *enemy = [[Enemy alloc] init];
    [self addChild:enemy];
    
    CGPoint position = CGPointMake(drand48() * self.size.width,
                                   drand48() * self.size.height);
    enemy.position = position;
    
    CGFloat angle = drand48() * M_PI;
    CGFloat magnitude = 20;
    enemy.physicsBody.velocity = CGVectorMake(cos(angle) * magnitude, sin(angle) * magnitude);
    
    return enemy;
}

- (void)didBeginContact:(SKPhysicsContact *)contact
{
    Enemy *enemy;
    if ([contact.bodyB.node isKindOfClass:[Enemy class]]) {
        enemy = (Enemy *)contact.bodyB.node;
    } else if ([contact.bodyA.node isKindOfClass:[Enemy class]]) {
        enemy = (Enemy *)contact.bodyA.node;
    }
    if (enemy && enemy.kind == self.player.kind) {
        [enemy removeFromParent];
        self.player.kind ++;
        self.player.kind = self.player.kind % 3;
        
        self.score ++;
        [self updateScore];
        
        [self.player updateColor];
    }
}

- (void)setupPlayer
{
    self.player = [[Player alloc] init];
    self.player.physicsBody.velocity = CGVectorMake(11.0, 5.0);
    [self addChild:self.player];
    self.player.position = CGPointMake(self.size.width / 2, self.size.height / 2);
}

- (void)setupScoreLabel
{
    self.scoreNode = [SKLabelNode labelNodeWithFontNamed:@"Helvetica Neue Thin"];
    self.scoreNode.zPosition = -1;
    self.scoreNode.fontSize = 65;
    self.scoreNode.position = CGPointMake(CGRectGetMidX(self.frame),
                                          CGRectGetMidY(self.frame));
    [self addChild:self.scoreNode];
    [self updateScore];
}

- (void)turn
{
    self.player.direction = (self.player.direction == kLeft ? kRight : kLeft);
}

- (void)updateScore
{
    self.scoreNode.text = [NSString stringWithFormat:@"%@", @(self.score)];
}

-(void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    [self turn];
}

- (void)update:(NSTimeInterval)currentTime
{
    CGVector velocity = self.player.physicsBody.velocity;
    CGFloat magnitude = sqrt(velocity.dx * velocity.dx + velocity.dy * velocity.dy);
    
    if (magnitude > 0) {
        CGVector normal = CGVectorMake(velocity.dx / magnitude, velocity.dy / magnitude);
        
        CGFloat angle = (M_PI / 20) * (self.player.direction == kLeft ? -1 : 1);
        normal.dx = normal.dx * cos(angle) - normal.dy * sin(angle);
        normal.dy = normal.dx * sin(angle) + normal.dy * cos(angle);
        
        CGVector force = CGVectorMake(normal.dx * 600.0, normal.dy * 600.0);
    
        [self.player.physicsBody applyForce:force];
    }
}

@end

File: Chain Pop/Chain Pop/GameScene.swift
//
//  GameScene.swift
//  Chain Pop
//
//  Created by Johannes Wärn on 05/12/14.
//  Copyright (c) 2014 Johannes Wärn. All rights reserved.
//

import SpriteKit

enum CollisionCategory : UInt32 {
    case Wall = 1
    case Bubble = 2
    case Bomb = 4
}

class GameScene: SKScene, SKPhysicsContactDelegate {
    override func didMoveToView(view: SKView) {
        physicsWorld.gravity = CGVectorMake(0.0, -0.1)
        physicsWorld.contactDelegate = self
        
        physicsBody = SKPhysicsBody(edgeLoopFromRect: view.frame)
        physicsBody?.categoryBitMask = CollisionCategory.Wall.rawValue
        physicsBody?.collisionBitMask = CollisionCategory.Bubble.rawValue
        physicsBody?.friction = 0.0
        physicsBody?.restitution = 1.0
        
        var noise = SKFieldNode.noiseFieldWithSmoothness(1.0, animationSpeed: 1.0)
        noise.strength = 0.0002
        self.addChild(noise)
        
        for n in 0 ..< 25 {
            addBubble()
        }
    }
    
    func addBubble(animated : Bool = false) {
        let bubble = Bubble()
        addChild(bubble)
        bubble.position = CGPoint(
            x: bubble.radius * 2 + (self.size.width  - bubble.radius) * CGFloat(randomDouble()),
            y: bubble.radius * 2 + (self.size.height - bubble.radius) * CGFloat(randomDouble())
        )
        if animated {
            bubble.xScale = 0
            bubble.yScale = bubble.xScale
            bubble.runAction(SKAction.scaleTo(1, duration: 0.5))
        }
    }
    
    override func touchesBegan(touches: NSSet, withEvent event: UIEvent) {
        for touch in touches.allObjects as [UITouch] {
            let bomb = Bomb()
            addChild(bomb)
            bomb.position = touch.locationInNode(self)
        }
    }
    
    func didBeginContact(contact: SKPhysicsContact) {
        if let bubble = contact.bodyA.node as? Bubble {
            if let bomb = contact.bodyB.node as? Bomb {
                let remove = SKAction.sequence([
                    SKAction.fadeAlphaTo(0, duration: 0.9),
                    SKAction.removeFromParent()
                    ])
                
                bomb.removeActionForKey("grow")
                
                if (bubble.kind == 0) {
                    self.runAction(SKAction.playSoundFileNamed("ljud3.caf", waitForCompletion: false))
                } else if (bubble.kind == 1) {
                    self.runAction(SKAction.playSoundFileNamed("ljud4.caf", waitForCompletion: false))
                } else if (bubble.kind == 2) {
                    self.runAction(SKAction.playSoundFileNamed("ljud5.caf", waitForCompletion: false))
                } else if (bubble.kind == 3) {
                    self.runAction(SKAction.playSoundFileNamed("ljud6.caf", waitForCompletion: false))
                } else {
                    self.runAction(SKAction.playSoundFileNamed("ljud7.caf", waitForCompletion: false))
                }
                bomb.runAction(remove)
            }
        }
    }
}

class Bubble: SKShapeNode {
    let radius: CGFloat = 5.0
    let possibleKinds = 5
    let kind: Int
    let color: SKColor
    
    override init() {
        kind = Int(arc4random()) % possibleKinds
        let hue = fmod(-0.04 + 1.0 / Double(possibleKinds) * Double(kind), 1.0 + randomDouble() * 0.6)
        let saturation = 0.58 + randomDouble() * 0.1
        let brightness = 0.75 + randomDouble() * 0.1
        color = SKColor(hue: CGFloat(hue), saturation: CGFloat(saturation), brightness: CGFloat(brightness), alpha: 1.0)
        
        super.init()
        path = CGPathCreateWithEllipseInRect(CGRectMake(-radius, -radius, radius * 2, radius * 2), nil)
        fillColor = color
        strokeColor = color
        name = "bubble"
        
        physicsBody = SKPhysicsBody(circleOfRadius: radius)
        physicsBody?.categoryBitMask = CollisionCategory.Bubble.rawValue
        physicsBody?.collisionBitMask = CollisionCategory.Bubble.rawValue | CollisionCategory.Wall.rawValue
        physicsBody?.contactTestBitMask = CollisionCategory.Bomb.rawValue
        physicsBody?.allowsRotation = false
        physicsBody?.linearDamping = 0.0
        physicsBody?.restitution = 1.0
        physicsBody?.friction = 0.0
        let angle = CGFloat(randomDouble() * M_PI)
        let magnitude = CGFloat(randomDouble() * 50 + 30)
        physicsBody?.velocity = CGVectorMake(cos(angle) * magnitude, sin(angle) * magnitude)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class Bomb: SKShapeNode {
    var kind: Int?
    var radius: CGFloat = 1.0
    var bubblesKilled : Int = 0
    
    override init() {
        super.init()
        path = CGPathCreateWithEllipseInRect(CGRectMake(-radius, -radius, radius * 2, radius * 2), nil)
        let color = SKColor(white: 1.0, alpha: 0.15)
        fillColor = color
        strokeColor = SKColor.clearColor()
        name = "bomb"
        
        physicsBody = SKPhysicsBody(circleOfRadius: radius)
        physicsBody?.categoryBitMask = CollisionCategory.Bomb.rawValue
        physicsBody?.collisionBitMask = 0
        physicsBody?.contactTestBitMask = 0
        physicsBody?.allowsRotation = false
        physicsBody?.pinned = true
        
        let grow = SKAction.runBlock {
            [unowned self] () -> () in
            self.radius += 1.0
            self.path = CGPathCreateWithEllipseInRect(CGRectMake(-self.radius, -self.radius, self.radius * 2, self.radius * 2), nil)
            self.physicsBody = SKPhysicsBody(circleOfRadius: self.radius)
            self.physicsBody?.categoryBitMask = CollisionCategory.Bomb.rawValue
            self.physicsBody?.collisionBitMask = 0
            self.physicsBody?.contactTestBitMask = 0
            self.physicsBody?.allowsRotation = false
            self.physicsBody?.pinned = true
        }
        runAction(SKAction.repeatActionForever(SKAction.sequence([SKAction.waitForDuration(1.0 / 60), grow])), withKey: "grow")
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

func randomDouble() -> Double {
    return Double(arc4random()) / Double(UINT32_MAX)
}